### ISSUE TITLE
nvme connect:  Failed to write to /dev/nvme-fabrics: Connection reset by peer

### ISSUE THREAD

**Issue by senyaroy** [spdk/spdk#697]
nvme connect over softRoCE is failing.
configured RDMA with softRoCE on ethernet NICs.
nvme discovery succeeded only once and is failing afterwards.

## Expected Behavior
Should connect

## Current Behavior
Failed to write to /dev/nvme-fabrics: Connection reset by peer

## Steps to Reproduce

1. ./scripts/rpc.py nvmf_subsystem_create -a -s SPDK00000000000000 nqn.2018-12.io.spdk:trgt204

2. ./scripts/rpc.py nvmf_subsystem_add_ns nqn.2018-12.io.spdk:trgt204 e9874d06-0502-4923-9c26-77598b48eb07

3. ./scripts/rpc.py nvmf_create_transport -t RDMA -u 8192 -p 4 -c 0

4. ./scripts/rpc.py nvmf_subsystem_add_listener -t RDMA -a <target ip> -s 4420 nqn.2018-12.io.spdk:trgt204

5. nvme discover -t rdma -a <target ip> -s 4420
Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2018-12.io.spdk:trgt204
traddr:  <target ip>
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000

6. nvme connect -t rdma -n "nqn.2018-12.io.spdk:trgt204" -a <target ip> -s 4420
Failed to write to /dev/nvme-fabrics: Connection reset by peer

7. nvme connect -t rdma -n nqn.2018-12.io.spdk:trgt204 -a <target ip> -s 4420
Failed to write to /dev/nvme-fabrics: Connection reset by peer

8. nvme discover -t rdma -a <target ip> -s 4420
Failed to write to /dev/nvme-fabrics: Connection reset by peer

9. dmesg | tail -20 (on initiator)
[  429.515103] RPC: Registered rdma transport module.
[  429.515107] RPC: Registered rdma backchannel transport module.
[  429.751598] Rounding down aligned max_sectors from 4294967295 to 4294967288
[ 8795.853470] rdma_rxe: already configured on eno1
[15666.576284] warning: `ethtool' uses legacy ethtool link settings API, link modes are only partially reported
[116397.973735] perf: interrupt took too long (2504 > 2500), lowering kernel.perf_event_max_sample_rate to 79000
[538031.487646] perf: interrupt took too long (3141 > 3130), lowering kernel.perf_event_max_sample_rate to 63000
[615636.006466] Ebtables v2.0 unregistered
[696289.544854] nvme nvme2: Connect rejected: status 8 (invalid service ID).
[696289.544945] nvme nvme2: rdma connection establishment failed (-104)
[697229.932267] nvme nvme2: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr <target ip>:4420
[697236.388469] nvme nvme2: Removing ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery"
[697309.333827] nvme nvme2: Connect rejected: status 10 (stale conn).
[697309.334531] nvme nvme2: rdma connection establishment failed (-104)
[697358.817175] nvme nvme2: Connect rejected: status 10 (stale conn).
[697358.817270] nvme nvme2: rdma connection establishment failed (-104)
[697412.945059] nvme nvme2: Connect rejected: status 10 (stale conn).
[697412.945153] nvme nvme2: rdma connection establishment failed (-104)
[697453.890792] nvme nvme2: Connect rejected: status 10 (stale conn).
[697453.890895] nvme nvme2: rdma connection establishment failed (-104)

10. lsmod | grep nvme
nvme_rdma              28408  0
nvme_fabrics           19997  1 nvme_rdma
nvmet_rdma             23175  0
nvmet                  51048  1 nvmet_rdma
rdma_cm                59673  7 rpcrdma,ib_srp,nvme_rdma,nvmet_rdma,ib_iser,rdma_ucm,ib_isert
ib_core               242235  17 rdma_cm,i40iw,ib_cm,iw_cm,rpcrdma,ib_srp,ib_ucm,nvme_rdma,nvmet_rdma,ib_iser,ib_srpt,ib_umad,ib_uverbs,rdma_rxe,rdma_ucm,ib_ipoib,ib_isert
nvme                   32338  0
nvme_core              58852  3 nvme,nvme_fabrics,nvme_rdma

11. ibv_rc_pingpong -g 0 -d rxe0 <target ip>
  local address:  LID 0x0000, QPN 0x000011, PSN 0xd14396, GID fe80::a6bf:1ff:fe06:9860
client read: Success
Couldn't read remote address

12. systemctl status firewalld
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Tue 2019-02-26 18:39:59 EST; 23h ago
     Docs: man:firewalld(1)
  Process: 28566 ExecStart=/usr/sbin/firewalld --nofork --nopid $FIREWALLD_ARGS (code=exited, status=0/SUCCESS)
 Main PID: 28566 (code=exited, status=0/SUCCESS)

13. Target:
ibv_rc_pingpong -d rxe0 -g 0
  local address:  LID 0x0000, QPN 0x000014, PSN 0x19cc6b, GID fe80::a6bf:1ff:fe06:2407
Failed to modify QP to RTR
Couldn't connect to remote QP

14. dmesg | tail -50 (on target)
[   17.758293] IPv6: ADDRCONF(NETDEV_UP): enp3s0f3: link is not ready
[   17.869696] IPv6: ADDRCONF(NETDEV_UP): enp3s0f3: link is not ready
[   17.869792] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[   17.872265] IPv6: ADDRCONF(NETDEV_UP): ens785f2: link is not ready
[   17.885765] IPv6: ADDRCONF(NETDEV_UP): ens785f2: link is not ready
[   17.888182] IPv6: ADDRCONF(NETDEV_UP): ens785f3: link is not ready
[   17.893153] Netfilter messages via NETLINK v0.30.
[   17.901418] IPv6: ADDRCONF(NETDEV_UP): ens785f3: link is not ready
[   17.902123] ip_set: protocol 6
[   21.420228] igb 0000:03:00.0 eno1: igb: eno1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX/TX
[   21.420375] IPv6: ADDRCONF(NETDEV_CHANGE): eno1: link becomes ready
[   72.927177] warning: `ethtool' uses legacy ethtool link settings API, link modes are only partially reported
[   73.412943] i40e: Registered client i40iw
[   73.714192] Rounding down aligned max_sectors from 4294967295 to 4294967288
[   73.746280] Loading iSCSI transport class v2.0-870.
[   73.758500] iscsi: registered transport (iser)
[   73.883200] RPC: Registered named UNIX socket transport module.
[   73.883204] RPC: Registered udp transport module.
[   73.883206] RPC: Registered tcp transport module.
[   73.883207] RPC: Registered tcp NFSv4.1 backchannel transport module.
[   73.897927] RPC: Registered rdma transport module.
[   73.897931] RPC: Registered rdma backchannel transport module.
[   80.744026] rdma_rxe: loaded
[   80.752382] rdma_rxe: set rxe0 active
[   80.752385] rdma_rxe: added rxe0 to eno1
[  286.077194] rdma_rxe: already configured on eno1
[63182.126381] rdma_rxe: already configured on eno1
[130136.251245] perf: interrupt took too long (2501 > 2500), lowering kernel.perf_event_max_sample_rate to 79000
[249114.876095] perf: interrupt took too long (3127 > 3126), lowering kernel.perf_event_max_sample_rate to 63000
[667945.126843] Ebtables v2.0 unregistered
[675895.301775] Generic UIO driver for PCI 2.3 devices version: 0.01.0
[675902.741088] ioatdma 0000:00:04.0: Removing dma and dca services
[675902.741194] igb 0000:03:00.0: DCA disabled
[675902.741223] igb 0000:03:00.3: DCA disabled
[675902.790197] ioatdma 0000:80:04.0: Removing dma and dca services
[675902.882159] ioatdma 0000:00:04.1: Removing dma and dca services
[675902.936711] ioatdma 0000:80:04.1: Removing dma and dca services
[675903.045277] ioatdma 0000:00:04.2: Removing dma and dca services
[675903.097739] ioatdma 0000:80:04.2: Removing dma and dca services
[675903.202970] ioatdma 0000:00:04.3: Removing dma and dca services
[675903.256901] ioatdma 0000:80:04.3: Removing dma and dca services
[675903.357673] ioatdma 0000:00:04.4: Removing dma and dca services
[675903.412356] ioatdma 0000:80:04.4: Removing dma and dca services
[675903.519053] ioatdma 0000:00:04.5: Removing dma and dca services
[675903.573434] ioatdma 0000:80:04.5: Removing dma and dca services
[675903.675036] ioatdma 0000:00:04.6: Removing dma and dca services
[675903.727160] ioatdma 0000:80:04.6: Removing dma and dca services
[675903.824394] ioatdma 0000:00:04.7: Removing dma and dca services
[675903.877696] ioatdma 0000:80:04.7: Removing dma and dca services
[690344.824485] Bits 55-60 of /proc/PID/pagemap entries are about to stop being page-shift some time soon. See the linux/Documentation/vm/pagemap.txt for details.

## Context (Environment including OS version, SPDK version, etc.)
SPDK: v19.01
OS: CentOS Linux release 7.6.1810 (Core)
Kernel: Linux 3.10.0-957.5.1.el7.x86_64
nvme version 1.6

**Comment by yidong0635** [comment]
@senyaroy , from your logs, I found that the firewall should be disabled.

**Comment by senyaroy** [comment]
@yidong0635 as you see step 12. firewall is disabled.

12\. systemctl status firewalld
     ● firewalld.service - firewalld - dynamic firewall daemon
     Loaded: loaded (/usr/lib/systemd/system/firewalld.service; enabled; vendor preset: enabled)
     Active: **inactive (dead)** since Tue 2019-02-26 18:39:59 EST; 23h ago

**Comment by yidong0635** [comment]
how about `systemctl disable firewalld` ?

**Comment by senyaroy** [comment]
systemctl disable firewalld
Removed symlink /etc/systemd/system/multi-user.target.wants/firewalld.service.
Removed symlink /etc/systemd/system/dbus-org.fedoraproject.FirewallD1.service.

systemctl status firewalld
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: enabled)
   Active: inactive (dead)
     Docs: man:firewalld(1)

nvme discover -t rdma -a <target ip> -s 4420
Failed to write to /dev/nvme-fabrics: Connection reset by peer

ibv_rc_pingpong -d rxe0 -g 1 <target ip>
  local address:  LID 0x0000, QPN 0x000011, PSN 0xfe1405, GID ::ffff:10.250.50.207
client read: Success
Couldn't read remote address

**Comment by yidong0635** [comment]
@senyaroy  I found that  your discovery
nvme discover -t rdma -a  -s 4420
    Discovery Log Number of Records 1, Generation counter 3
    =====Discovery Log Entry 0======
    trtype:  rdma
    **_adrfam:  ipv4_**
    subtype: nvme subsystem
    treq:    not specified
    portid:  0
    trsvcid: 4420

dmesg | tail -50 (on target)  IPV6 log message.

please to keep the the same when creating and connecting.  some creation commands with parameter
_-f IPV6._

**Comment by yidong0635** [comment]
follow your steps, I can create it in my host using soft-roce with filling out IP address. -a 192.168.xx . 
you can use `./scripts/rpc.py xxx -h` to check if need some parameters.

**Comment by yidong0635** [comment]
[root@waikikibeach13 spdk]# nvme discover -t rdma -a 10.67.111.142 -s 4420

Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2018-12.io.spdk:trgt204
traddr:  10.67.111.142
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000
[root@waikikibeach13 spdk]# nvme connect -t rdma -n "nqn.2018-12.io.spdk:trgt204" -a 10.67.111.142 -s 4420
[root@waikikibeach13 spdk]#

**Comment by yidong0635** [comment]
confirm your target is still alive.

**Comment by senyaroy** [comment]
yes my target is still alive.
it is ipv4 what i am using.
i discover with target ip address and it succeeds only once,
from next time it gives the following error.
nvme discover -t rdma -a <target_ip> -s 4420
Failed to write to /dev/nvme-fabrics: Connection reset by peer

**Comment by yidong0635** [comment]
root@waikikibeach20:/home/yidong/spdk# nvme discover -t rdma -a 192.168.1.1 -s 4420
Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2018-12.io.spdk:trgt204
traddr:  192.168.1.1
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000
root@waikikibeach20:/home/yidong/spdk# nvme discover -t rdma -a 192.168.1.1 -s 4420

Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2018-12.io.spdk:trgt204
traddr:  192.168.1.1
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000
root@waikikibeach20:/home/yidong/spdk# nvme connect -t rdma -n "nqn.2018-12.io.spdk:trgt204" -a 192.168.1.1 -s 4420
root@waikikibeach20:/home/yidong/spdk# nvme discover -t rdma -a 192.168.1.1 -s 4420
Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2018-12.io.spdk:trgt204
traddr:  192.168.1.1
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000
root@waikikibeach20:/home/yidong/spdk# nvme connect -t rdma -n "nqn.2018-12.io.spdk:trgt204" -a 192.168.1.1 -s 4420
Failed to write to /dev/nvme-fabrics: Operation already in progress
root@waikikibeach20:/home/yidong/spdk# nvme disconnect -h
Usage: nvme disconnect <device> [OPTIONS]

Disconnect from NVMeoF subsystem

Options:
  [  --nqn=<LIST>, -n <LIST> ]          --- nqn name
  [  --device=<LIST>, -d <LIST> ]       --- nvme device
root@waikikibeach20:/home/yidong/spdk# nvme disconnect -n "nqn.2018-12.io.spdk:trgt204"
NQN:nqn.2018-12.io.spdk:trgt204 disconnected 1 controller(s)
root@waikikibeach20:/home/yidong/spdk# nvme connect -t rdma -n "nqn.2018-12.io.spdk:trgt204" -a 192.168.1.1 -s 4420
root@waikikibeach20:/home/yidong/spdk#

**Comment by changpe1** [comment]
@senyaroy softRocE only supports limited queue pair connections, so you can try use the parameter --nr-io-queues=1 with your nvme connect command, initiator will use number of cores for IO queues by default.

**Comment by revanbiradar2019** [comment]
Discovery is only failing . So we cannot go for connecting,

**Comment by revanbiradar2019** [comment]
Hi We tried the following. Still connecting error is found

[‎3/‎7/‎2019 5:24 PM]  Devanathan Karunakaran (Industrial & Engineering Services (I&ES)):  
No Title 
root@testubuntu3:/home/ubuntu# nvme discover -t rdma -a 10.201.40.37 -s 4420

Discovery Log Number of Records 2, Generation counter 5
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2016-06.io.spdk:5node37
traddr:  10.201.40.37
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000
=====Discovery Log Entry 1======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  1
trsvcid: 4420
subnqn:  nqn.2016-06.io.spdk:570node37
traddr:  10.201.40.37
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000
root@testubuntu3:/home/ubuntu# nvme connect -t rdma -n "nqn.2016-06.io.spdk:5node37" -a 10.201.40.37 -s 4420 --nr-io-queues=1
Failed to write to /dev/nvme-fabrics: Connection reset by peer

**Comment by darsto** [comment]
@Seth5141 to take a look

**Comment by Smalinuxer** [comment]
i retest it in spdk [v19.01](https://github.com/spdk/spdk/commit/e3e35a85ea69dc101507b9ff9838074c55fc7a40) , but not get any error .

steps : 
```
./scripts/rpc.py nvmf_subsystem_create -a  -s SPDK00000000000001 nqn.2016-06.io.spdk:cnode1
./scripts/rpc.py nvmf_subsystem_add_ns nqn.2016-06.io.spdk:cnode1 Nvme0n1
./scripts/rpc.py nvmf_create_transport -t RDMA -u 8192 -p 4 -c 0
./scripts/rpc.py nvmf_subsystem_add_listener -t RDMA -a 192.168.91.11 -s 4420 nqn.2016-06.io.spdk:cnode1
```
and then cmd
```

root@workspace:/home/jiaqizho/spdk-issue697# nvme discover -t rdma -a 192.168.91.11 -s 4420

Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2016-06.io.spdk:cnode1
traddr:  192.168.91.11
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000

root@workspace:/home/jiaqizho/spdk-issue697# nvme list
Node             SN                   Model                                    Namespace Usage                      Format           FW Rev
---------------- -------------------- ---------------------------------------- --------- -------------------------- ---------------- --------
/dev/nvme0n1     SPDK00000000000001   SPDK bdev Controller                     1         800.17  GB / 800.17  GB    512   B +  0 B   19.01

```

it seem like work good . 

can you put yuor nvmf_tgt config and nvmf_tgt output when occur "Failed to write to /dev/nvme-fabrics: Connection reset by peer" ?

**Comment by wanqunintel** [comment]
@revanbiradar2019 , i tried to reproduce your issue, the steps are as follows
1. ./scripts/rpc.py nvmf_subsystem_create -a -s SPDK00000000000000 nqn.2018-12.io.spdk:trgt204
2. ./scripts/rpc.py nvmf_subsystem_add_ns nqn.2018-12.io.spdk:trgt204 e9874d06-0502-4923-9c26-77598b48eb07
Exception:
Got JSON-RPC error response
response:
{
  "code": -32602,
  "message": "Invalid parameters"
}
3. ./scripts/rpc.py nvmf_create_transport -t RDMA -u 8192 -p 4 -c 0
4. **./scripts/rpc.py nvmf_subsystem_add_listener -t RDMA -a -s 4420 nqn.2018-12.io.spdk:trgt204
usage: rpc.py nvmf_subsystem_add_listener [-h] -t TRTYPE -a TRADDR [-f ADRFAM]
                                          [-s TRSVCID]
                                          nqn
rpc.py nvmf_subsystem_add_listener: error: argument -a/--traddr: expected one argument**
It seems step 4 has issues. after i use following command with IP address added, it passed
./scripts/rpc.py nvmf_subsystem_add_listener -t RDMA -a 172.28.128.11 -s 4420 nqn.2018-12.io.spdk:trgt204
5. nvme discover -t rdma -a 172.28.128.11 -s 4420

Discovery Log Number of Records 1, Generation counter 3
=====Discovery Log Entry 0======
trtype:  rdma
adrfam:  ipv4
subtype: nvme subsystem
treq:    not specified
portid:  0
trsvcid: 4420
subnqn:  nqn.2018-12.io.spdk:trgt204
traddr:  172.28.128.11
rdma_prtype: not specified
rdma_qptype: connected
rdma_cms:    rdma-cm
rdma_pkey: 0x0000

So you need to define the IP address when add listener and also add IP address when use nvme discover.
Would you have a try with above steps by assigned IP address to two commands to check if it existed
./scripts/rpc.py nvmf_subsystem_add_listener -t RDMA -a 172.28.128.11 -s 4420 nqn.2018-12.io.spdk:trgt204
nvme discover -t rdma -a 172.28.128.11 -s 4420

commit id e208be6f1d52c25c7096053efd018fc525407fda

**Comment by changpe1** [comment]
@senyaroy as we can't reproduce the issue, so if no other input, I'm going to close the issue, thanks.

**Comment by senyaroy** [comment]
Thanks All for the responses.
The basic thing to be noted here is that we were facing this issue on a softRoCE based setup.
Otherwise, we now have an RDMA capable NIC and it is working fine.

**Comment by yidong0635** [comment]
@senyaroy  please show us :
1. how you setup your softroce, use command to install or compile source code by yourself ?
2. please use `rxe_cfg status` to list your softroce environments.

**Comment by optimistyzy** [comment]
I think that it is third party issue with software roce, I will close this issue. If there are still relevant issues, please open it again.

**Comment by jainsmit** [comment]
I was also seeing similar issue and later able to resolve this issue by configuring the right NvME RoCE Interface

So here is the background of my issue. I had two interfaces but NvME RoCE interface was not configured (Which I was not aware). FA was reachable from already configured interface and was seeing below issue:

`root@dev-quill-touch-0:~# nvme connect -t rdma -n "nqn.2010-06.com.purestorage:flasharray.124aa41219ec353b" -a 192.168.1.130 -s 4420
Failed to write to /dev/nvme-fabrics: Connection reset by peer
root@dev-quill-touch-0:~# `

Dmesg was also showing similar error as mentioned in this thread:

`[Mon Dec 26 05:29:47 2022] Command line: BOOT_IMAGE=/boot/vmlinuz-5.4.0-107-generic
root=UUID=a594bda9-1e75-41f4-a95e-6b3d836803de ro maybe-ubiquity net.ifnames=0
biosdevname=0 ipv6.disable=1 nvme_core.multipath=N

[Mon Dec 26 05:29:47 2022] Kernel command line: BOOT_IMAGE=/boot/vmlinuz-5.4.0-107-generic
root=UUID=a594bda9-1e75-41f4-a95e-6b3d836803de ro maybe-ubiquity net.ifnames=0 biosdevname=0 ipv6.disable=1 nvme_core.multipath=N
[Mon Dec 26 06:14:36 2022] nvme nvme0: rdma connection establishment failed (-104)

[Mon Dec 26 06:14:36 2022] nvme nvme0: rdma connection establishment failed (-104)

[Mon Dec 26 06:14:36 2022] nvme nvme0: rdma connection establishment failed (-104)

[Mon Dec 26 06:14:40 2022] nvme nvme0: rdma connection establishment failed (-104)


<b>Later I found from kernel log  in a node that `mlx5_core` interface is down and that gave me the clue. Once I configured the `mlx5_core` which was the right interface for NvME RoCE for a client, this issue got resolved.</b>

`Dec 26 22:19:29 ubuntu kernel: [    8.447146] mlx5_core 0000:03:00.0 eth3: Link up `
